packages/cli/src/lib/version.ts

```diff
@@ -60,3 +62,22 @@ export function findVersion() {
 
 export const version = findVersion();
 export const isDev = fs.pathExistsSync(paths.resolveOwn('src'));
+
+export function createPackageVersionProvider(lockfile?: Lockfile) {
+  return (name: string, versionHint?: string) => {
+    const targetVersion = versionHint || packageVersions[name];
+    if (!targetVersion) {
+      throw new Error(`No version available for package ${name}`);
+    }
+
+    const lockfileEntries = lockfile?.get(name);
+    if (lockfileEntries?.some(entry => entry.range === '*')) {
+      return '*';
+    }
+    const validRanges = lockfileEntries?.filter(entry =>
+      semver.satisfies(targetVersion, entry.range),
+    );
+    const highestRange = validRanges?.slice(-1)[0];
```

Comment:
are they really guaranteed to be in order?

Reply:
Yeah the lockfile places them in a nice order. The `semver` package doesn't have any range sorting or anything like that so I cba diving into that